[memprof] Add CallStackRadixTreeBuilder

kazutakahirata opened the pull request:
Call stacks are a huge portion of the MemProf profile, taking up 70+%
of the profile file size.

This patch implements a radix tree to compress call stacks, which are
known to have long common prefixes.  Specifically,
CallStackRadixTreeBuilder, introduced in this patch, takes call stacks
in the MemProf profile, sorts them in the dictionary order to maximize
the common prefix between adjacent call stacks, and then encodes a
radix tree into a single array that is ready for serialization.

The resulting radix array is essentially a concatenation of call stack
arrays, each encoded with its length followed by the payload, except
that these arrays contain "instructions" like "skip 7 elements
forward" to borrow common prefixes from other call stacks.

This patch does not integrate with the MemProf
serialization/deserialization infrastructure yet.  Once integrated,
the radix tree is expected to roughly halve the file size of the
MemProf profile.

kazutakahirata committed:
[memprof] Add CallStackRadixTreeBuilder

Call stacks are a huge portion of the MemProf profile, taking up 70+%
of the profile file size.

This patch implements a radix tree to compress call stacks, which are
known to have long common prefixes.  Specifically,
CallStackRadixTreeBuilder, introduced in this patch, takes call stacks
in the MemProf profile, sorts them in the dictionary order to maximize
the common prefix between adjacent call stacks, and then encodes a
radix tree into a single array that is ready for serialization.

The resulting radix array is essentially a concatenation of call stack
arrays, each encoded with its length followed by the payload, except
that these arrays contain "instructions" like "skip 7 elements
forward" to borrow common prefixes from other call stacks.

This patch does not integrate with the MemProf
serialization/deserialization infrastructure yet.  Once integrated,
the radix tree is expected to roughly halve the file size of the
MemProf profile.

teresajohnson commented on llvm/lib/ProfileData/MemProf.cpp:
```diff
@@ -410,6 +410,161 @@ CallStackId hashCallStack(ArrayRef<FrameId> CS) {
+  // Create a list of call stacks to be sorted.
+  std::vector<CSIdPair> CallStacks;
+  CallStacks.reserve(MemProfCallStackData.size());
+  for (auto &[CSId, CallStack] : MemProfCallStackData) {
```
If we expect MemProfCallStackData to be unused by the caller of build(), could it be std::move'd and then use takeVector here to avoid the copying?

kazutakahirata replied:
This is really a good idea!  After I switched to `takeVector()`, `sortCallStacks` doesn't have much going on, so I've folded it into its caller -- `build()`.

teresajohnson commented on llvm/lib/ProfileData/MemProf.cpp:
```diff
@@ -410,6 +410,161 @@ CallStackId hashCallStack(ArrayRef<FrameId> CS) {
+// into RadixArray.  If a given call stack doesn't have a common prefix relative
+// to the previous one, we emit:
+//
+// - the frames in the given call stack in the reverse order
```
by "reverse" do you mean "leaf to root" order? Prefer to be more explicit here.

teresajohnson replied:
Oh, skimming through the rest of the code I see you actually mean that you are produce a reversed RadixArray and later reverse it in place. I think rather than putting the reverse order note here, this deserves a more explicit comment at the start of this description.

kazutakahirata replied:
Expanded comments.

teresajohnson commented on llvm/lib/ProfileData/MemProf.cpp:
```diff
@@ -410,6 +410,161 @@ CallStackId hashCallStack(ArrayRef<FrameId> CS) {
+    const llvm::SmallVector<FrameId> *CallStack,
+    const llvm::SmallVector<FrameId> *Prev,
+    const llvm::DenseMap<FrameId, uint32_t> &MemProfFrameIndexes) {
+  // Compute the length of the common prefix between Prev and CallStack.
```
maybe "common root prefix"?

kazutakahirata replied:
Done.

teresajohnson reviewed (commented):
Started a review, but am confused about what "Indexes" is - can you clarify that with a more intuitive description? I added a few other comments too but will finish the review once I have a better understanding of that vector.

kazutakahirata committed:
Expand comments.

Call MemProfCallStackData.takeVector() to take advantage of the move
semantics.

kazutakahirata commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -905,6 +905,106 @@ struct IndexedMemProfData {
+// we just need a pointer from a given call stack to its parent.  For example,
+// CallStackId 2 would point to CallStackId 1 as a parent.
+//
+// We serialize the radix tree above into a single array along with the length
```
Done.

kazutakahirata commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -905,6 +905,106 @@ struct IndexedMemProfData {
+};
+
+llvm::DenseMap<CallStackId, uint32_t>
+writeCallStackRadixTree(raw_ostream &OS,
```
Sorry, this function declaration doesn't have its corresponding body.  I've deleted this in the latest iteration.

kazutakahirata commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -905,6 +905,106 @@ struct IndexedMemProfData {
+  // Mapping from CallStackIds to indexes into RadixArray.
+  llvm::DenseMap<CallStackId, uint32_t> CallStackPos;
+
+  // The indexes within RadixArray of the last call stack's frames encoded
```
I've expanded comments in the latest iteration.

kazutakahirata commented:
Thank you so much for reviewing this PR!  I hope the comments about `Indexes` make more sense now.  Please take a look again.  Thanks!

kazutakahirata committed:
Merge branch 'main' into memprof_v3_radix_tree

kazutakahirata committed:
Use LinearFrameId and LinearCallStackId.

kazutakahirata committed:
Use LinearFrameId and LinearCallStackId in more places.

teresajohnson commented on llvm/lib/ProfileData/MemProf.cpp:
```diff
@@ -410,6 +410,153 @@ CallStackId hashCallStack(ArrayRef<FrameId> CS) {
+    Prev = &CallStack;
+  }
+
+  if (RadixArray.size() >= 2) {
```
What happens if RadixArray.size() < 2? If it is 0, is that a case where we could/should be returning early from this function? If 1, afaict the loops below will end up as no-ops (first shouldn't execute, second I believe will just set 0 to -0 if I am following how the CallStackPos would be set up in that case).

kazutakahirata replied:
I added `RadixArray.size() >= 2` because I ran into a problem handling the empty `RadixArray` exposed by `TEST(MemProf, RadixTreeBuilderEmpty)`.  Given the empty `RadixArray`, `I` and `J` are set to `0` and `SIZE_MAX`, respectively, because `-1` is treated as `SIZE_MAX`.

I'll get rid of this `if` statement.  In exchange, I'll return early in case of the empty input.  Then put `assert(!RadixArray.empty())` just before the in-place array reversal code to guard against the `SIZE_MAX` problem.

teresajohnson reviewed (approved):
lgtm with one question/possible refinement below

kazutakahirata committed:
Teach CallStackRadixTreeBuilder::build to return early.
